src/ore/src/region.rs

```diff
@@ -148,3 +150,232 @@ impl<T> LgAllocRegion<T> {
+}
+
+impl<T> MMapRegion<T> {
+    /// Clear the contents of this region without dropping elements.
+    unsafe fn clear(&mut self) {
+        self.inner.set_len(0);
+    }
+}
+
+impl<T: Debug> Debug for MMapRegion<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MMapRegion")
+            .field("inner", &self.inner)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<T> Deref for MMapRegion<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> Default for Region<T> {
+    #[inline]
+    fn default() -> Self {
+        Self::new_empty()
+    }
+}
+
+impl<T> Region<T> {
+    /// Create a new empty region.
+    #[inline]
+    #[must_use]
+    pub fn new_empty() -> Region<T> {
+        Region::Heap(Vec::new())
+    }
+
+    /// Create a new heap-allocated region of a specific capacity.
+    #[inline]
+    #[must_use]
+    pub fn new_heap(capacity: usize) -> Region<T> {
+        Region::Heap(Vec::with_capacity(capacity))
+    }
+
+    /// Create a new file-based mapped region of a specific capacity. The capacity of the
+    /// returned region can be larger than requested to accommodate page sizes.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the memory allocation fails.
+    #[inline(always)]
+    pub fn new_mmap(capacity: usize) -> Result<Region<T>, lgalloc::AllocError> {
+        lgalloc::allocate(capacity).map(|(ptr, capacity, handle)| {
+            // SAFETY: memory points to suitable memory.
+            // It is UB to call `from_raw_parts` with a pointer not allocated from the system
+            // allocator, but we accept this here because we promise never to reallocate the vector.
+            let inner =
+                ManuallyDrop::new(unsafe { Vec::from_raw_parts(ptr.as_ptr(), 0, capacity) });
+            let handle = Some(handle);
+            Region::MMap(MMapRegion { inner, handle })
+        })
+    }
+
+    /// Create a region depending on the capacity.
+    ///
+    /// The capacity of the returned region must be at least as large as the requested capacity,
+    /// but can be larger if the implementation requires it.
+    ///
+    /// Returns a [`Region::MMapRegion`] if possible, and falls back to [`Region::Heap`] otherwise.
+    #[must_use]
+    pub fn new_auto(capacity: usize) -> Region<T> {
+        match Region::new_mmap(capacity) {
+            Ok(r) => return r,
+            Err(lgalloc::AllocError::Disabled) | Err(lgalloc::AllocError::InvalidSizeClass(_)) => {}
+            Err(e) => {
+                eprintln!("lgalloc error: {e}, falling back to heap");
+            }
+        }
+        // Fall-through
+        Region::new_heap(capacity)
+    }
+
+    /// Clears the contents of the region, without dropping its elements.
+    ///
+    /// # Safety
+    ///
+    /// Discards all contends. Elements are not dropped.
+    #[inline]
+    pub unsafe fn clear(&mut self) {
```

Comment:
Why is this method `unsafe`? Looking at the docs, `vec.set_len(0)` should always be safe for any vec.

Reply:
I guess you could argue that we want to alert callers to the fact that this can leak memory. That's not a safety issue, but perhaps still good to flag.